### ISSUE TITLE
[FEAT] Better Storage for ITEMs

### ISSUE THREAD

**Issue by iceBear67** [InlinedLambdas/AstralFlow#42]
a reliable and compat storage for items.

And there are my ideas

1. Write data into items directly
2. Pushing extra data into item uuids

The first way seems easiest and most benefit. By this way we can dispose unused items and reduce our work to design such a structure. 

The second way might be effective by pushing file index into uuid. However it is not compat because it will produce tons of unused spaces by the time.

We should combine their advantage to solve these following problems.

1. *Lifecycle.*  Plugins cant monitor that some item jist disposed. And if we dont deal with that we will ran out the disk at last
2. *Compat.* We need to minimalize the usage
3. *Effective* Cant be too slow
4. *Reliable.* Security is also important

**Comment by iceBear67** [comment]
It is ineffectient about filtering the pdc from itemstack.

I will add an option about filter,and AF will get its first magics.

1. Filtering over packets.
2. JDK Redefine Magics.
3. Server patching

**Comment by iceBear67** [comment]
I found where we should patch.  

At `FriendlyByteBuf` (Paper, called `PacketDataSerializer` in spigot)
```java
    public FriendlyByteBuf writeItem(ItemStack stack) {
        if (stack.isEmpty() || stack.getItem() == null) { // CraftBukkit - NPE fix itemstack.getItem()
            this.writeBoolean(false);
        } else {
            this.writeBoolean(true);
            Item item = stack.getItem();

            this.writeVarInt(Item.getId(item));
            this.writeByte(stack.getCount());
            CompoundTag nbttagcompound = null;

            if (item.canBeDepleted() || item.shouldOverrideMultiplayerNbt()) {
                // Spigot start - filter
                stack = stack.copy();
                // CraftItemStack.setItemMeta(stack, CraftItemStack.getItemMeta(stack)); // Paper - This is no longer needed due to NBT being supported
                // Spigot end
                nbttagcompound = stack.getTag();
                // Paper start
                if (nbttagcompound != null && nbttagcompound.contains("SkullOwner", 10)) {
                    CompoundTag owner = nbttagcompound.getCompound("SkullOwner");
                    if (owner.hasUUID("Id")) {
                        nbttagcompound.putUUID("SkullOwnerOrig", owner.getUUID("Id"));
                        net.minecraft.world.level.block.entity.SkullBlockEntity.sanitizeUUID(owner);
                    }
                }
                // Paper end
            }

            this.writeNbt(nbttagcompound);
        }

        return this;
    }
```

Inserting hooks above `this.writeNbt(nbttagcompound)` would work, however it is leading to full of asm codes.

**Comment by iceBear67** [comment]
So here're our tasks:

1. Find CompoundTag in obfuscated environments. (including 1.18+, treat them as obfuscated)
2. Find FriendlyByteBuf / PacketDataSerializer in obfuscated environments.
3. Determine injector solution. Predefine or redefine or server patching? I prefer redefine with mixin but there're no projects has implemented mixin injection for bukkit ( or poor API )

**Comment by iceBear67** [comment]
This Optimization (nbt filter) will be separated into another module and introduce as an option.

**Comment by iceBear67** [comment]
This Optimization (nbt filter) will be separated into another module.  

#79

**Comment by iceBear67** [comment]
This depends on #79
